<a href="https://colab.research.google.com/github/krithusrinivasan/EDA-Assignments/blob/main/Dry_Bean_data_set_Logistic_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###                                 DRY BEAN DATASET

The Dry Bean dataset is a collection of data used to teach computers how to recognize different types of beans. Each type of bean has its own special shape and size, and the dataset includes these details for thousands of beans.

The dataset has measurements like:
Area: How big the bean is.
Perimeter: The length around the edge of the bean.
Shape Details: Things like how round the bean is or how stretched out it looks.

Significance:
1. Helping Farmers: By teaching a computer to recognize different beans, farmers can sort their beans faster and make sure they’re selling the right types.
2. Learning Machine Learning: This dataset is great for people who want to learn how to build models that can classify or sort things. It’s like practice material for new data scientists.
3. Building Machines: The data can also be used to design machines that automatically sort beans, making the process quicker and more accurate.

In [ ]:
#import libraries
import pandas as pd #for using data frames
from sklearn.linear_model import LogisticRegression #for using logistic regression
from sklearn.model_selection import train_test_split #for splitting the data into training and testing sets
from sklearn.metrics import accuracy_score #for calculating the accuracy of the model
from sklearn.metrics import confusion_matrix, classification_report #for evaluating the model

In [ ]:
#load dataset
data=pd.read_csv('/content/Dry_Bean_Dataset.csv')

In [ ]:
data.head()

Area  Perimeter  MajorAxisLength  MinorAxisLength  AspectRation  \
0  28395    610.291       208.178117       173.888747      1.197191   
1  28734    638.018       200.524796       182.734419      1.097356   
2  29380    624.110       212.826130       175.931143      1.209713   
3  30008    645.884       210.557999       182.516516      1.153638   
4  30140    620.134       201.847882       190.279279      1.060798   

   Eccentricity  ConvexArea  EquivDiameter    Extent  Solidity  roundness  \
0      0.549812       28715     190.141097  0.763923  0.988856   0.958027   
1      0.411785       29172     191.272751  0.783968  0.984986   0.887034   
2      0.562727       29690     193.410904  0.778113  0.989559   0.947849   
3      0.498616       30724     195.467062  0.782681  0.976696   0.903936   
4      0.333680       30417     195.896503  0.773098  0.990893   0.984877   

   Compactness  ShapeFactor1  ShapeFactor2  ShapeFactor3  ShapeFactor4  Class  
0     0.913358      0.007332      0.003147      0.834222      0.998724  SEKER  
1     0.953861      0.006979      0.003564      0.909851      0.998430  SEKER  
2     0.908774      0.007244      0.003048      0.825871      0.999066  SEKER  
3     0.928329      0.007017      0.003215      0.861794      0.994199  SEKER  
4     0.970516      0.006697      0.003665      0.941900      0.999166  SEKER

Head id used to get a quick look at the first few rows of a dataset. data.head() is used to preview the data, to check column names, to understand data types,to spot inital issues.

In [ ]:
data.shape
# To determine the dimensions of the dataset. Understanding the datasize, Planning Analysis

(13611, 17)

In [ ]:
print(data.columns)
# To understand the structure of the dataset. Getting column names.

Index(['Area', 'Perimeter', 'MajorAxisLength', 'MinorAxisLength',
       'AspectRation', 'Eccentricity', 'ConvexArea', 'EquivDiameter', 'Extent',
       'Solidity', 'roundness', 'Compactness', 'ShapeFactor1', 'ShapeFactor2',
       'ShapeFactor3', 'ShapeFactor4', 'Class'],
      dtype='object')


In [ ]:
# defining features and target variable(dropping the class variable because class is the target variable)
x= data.drop(columns=['Class'], inplace=False, axis=1)
y= data['Class']

In [ ]:
x.shape


(13611, 16)

In [ ]:
y.shape

(13611,)

In [ ]:
x.isnull().sum()
# To detect Missing or Null values in the dataset.

Area               0
Perimeter          0
MajorAxisLength    0
MinorAxisLength    0
AspectRation       0
Eccentricity       0
ConvexArea         0
EquivDiameter      0
Extent             0
Solidity           0
roundness          0
Compactness        0
ShapeFactor1       0
ShapeFactor2       0
ShapeFactor3       0
ShapeFactor4       0
dtype: int64

In [ ]:
y.isnull().sum()

0

In [ ]:
# split the dataset into training and testing sets and dividing data input and output variable
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

In [ ]:
#Train the Logistic regression model
classifier = LogisticRegression()
classifier.fit(x_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [ ]:
# Predict for x dataset
y_pred = classifier.predict(x_test)


In [ ]:
y_pred_df = pd.DataFrame({"actual": y_pred, "predicted" : classifier.predict(x_test)})


In [ ]:
y_pred_df.head(10)

actual predicted
0     SEKER     SEKER
1     HOROZ     HOROZ
2     SEKER     SEKER
3  DERMASON  DERMASON
4  DERMASON  DERMASON
5     SEKER     SEKER
6      CALI      CALI
7      SIRA      SIRA
8    BOMBAY    BOMBAY
9  DERMASON  DERMASON

In [ ]:
y_pred_df['correct'] = y_pred_df['actual'] == y_pred_df['predicted']
y_pred_df['correct'].mean()

1.0

In [ ]:
# Confusion Matrix for the model accuracy
confusion_matrix = confusion_matrix(y_test,y_pred)
print (confusion_matrix)


[[216   0 116   0  58   0   5]
 [  0 161   0   0   0   0   0]
 [114   0 340   0  17   5   3]
 [  0   0   0 895   6  61  81]
 [  6   0  27  20 348   8 179]
 [  6   0   0 132   9 390  82]
 [  0   0   0  56  96 116 531]]


1. Class 2 seems to be very well classified since it has 161 correct predictions and no misclassifications.
2. Class 1 and Class 3 have some confusion between them since there are significant misclassifications (116 from Class 1 to Class 3 and 114 from Class 3 to Class 1).
3. Classes 4, 5, 6, and 7 show some degree of confusion with each other.

In [ ]:
#classification report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

    BARBUNYA       0.63      0.55      0.59       395
      BOMBAY       1.00      1.00      1.00       161
        CALI       0.70      0.71      0.71       479
    DERMASON       0.81      0.86      0.83      1043
       HOROZ       0.65      0.59      0.62       588
       SEKER       0.67      0.63      0.65       619
        SIRA       0.60      0.66      0.63       799

    accuracy                           0.71      4084
   macro avg       0.72      0.71      0.72      4084
weighted avg       0.70      0.71      0.70      4084



1. BOMBAY has perfect precision, recall, and F1-score, indicating the model handles this class very well.
2. DERMASON also performs well, with high precision and recall.
3. BARBUNYA, HOROZ, and SIRA have lower scores, suggesting that the model struggles more with these classes.



1. Accuracy: The proportion of correctly predicted labels out of the total number of predictions.
2. Precision: The proportion of true positive predictions out of all positive predictions. High precision means that fewer positive results are actually false positives.
3. Recall: The proportion of true positive predictions out of all actual positives. High recall means that most actual positives are correctly identified.